src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp

```diff
@@ -0,0 +1,163 @@
+#include <cmath>
+#include <cstring>
+#include <limits>
+#include <pmmintrin.h>
+
+#define UNUSED(x) (void)(x)
+#define DEBUG_ONLY(x) (void)(x)
+
+#ifdef COMPILER_GCC
+
+#include "UnixSal.h"
+#define EXPORT_API(ret) extern "C" __attribute__((visibility("default"))) ret
+
+#else
+#include <intrin.h>
+#define EXPORT_API(ret) extern "C" __declspec(dllexport) ret __stdcall
+#endif
+
+EXPORT_API(void) CalculateIntermediateVariablesNative(int fieldCount, int latentDim, int count, _In_ int * fieldIndices, _In_ int * featureIndices, _In_ float * featureValues,
```

Comment:
>CalculateIntermediateVariables [](start = 17, length = 30)

Any chance you can add comments to this code? @wschin

Reply:
The two functions are Algorithm 1 & 2  at [https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf](https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf).

---
In reply to: [196861456](https://github.com/dotnet/machinelearning/pull/383#discussion_r196861456) [](ancestors = 196861456)